pkg/quota/image/imagestreamimport_evaluator.go

```diff
@@ -0,0 +1,220 @@
+
+	for _, imageSpec := range isi.Spec.Images {
+		if imageSpec.From.Kind != "DockerImage" {
+			continue
+		}
+		ref, err := c.GetImageReferenceForObjectReference(isi.Namespace, &imageSpec.From)
+		if err != nil {
+
+		}
+		if !c.processedSpecRefs.Has(ref) {
+			c.processedSpecRefs.Insert(ref)
+			specRefsIncrement.Set(specRefsIncrement.Value() + 1)
+		}
+		if !c.processedRefs.Has(ref) {
+			c.processedRefs.Insert(ref)
+			allRefsIncrement.Set(allRefsIncrement.Value() + 1)
+		}
+	}
+
+	return map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *specRefsIncrement,
+		imageapi.ResourceImageStreamImages: *allRefsIncrement,
+	}
+}
+
+// getRepositoryUsageIncrement returns a usage increment for ImageStreamImport.Spec.Repository. The repository
+// causes image importer to import all the repository tags in an image stream. We don't attempt to resolve
+// number of tags here. We just expect the worst case scenario where the max limit of import tags per
+// repository is reached and decrease it by the number of image references pointing to the repository that are
+// already tagged in a project.
+func (c *imageStreamImportUsageComputer) getRepositoryUsageIncrement(isi *imageapi.ImageStreamImport) kapi.ResourceList {
+	specRefsIncrement := resource.NewQuantity(0, resource.DecimalSI)
+	allRefsIncrement := resource.NewQuantity(0, resource.DecimalSI)
+	usage := map[kapi.ResourceName]resource.Quantity{
+		imageapi.ResourceImageStreamTags:   *specRefsIncrement,
+		imageapi.ResourceImageStreamImages: *allRefsIncrement,
+	}
+
+	if isi.Spec.Repository == nil || isi.Spec.Repository.From.Kind != "DockerImage" {
+		return usage
+	}
+
+	repoRef, err := c.GetImageReferenceForObjectReference(isi.Namespace, &isi.Spec.Repository.From)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to resolve Spec.Repository.From of isimport %s/%s: %v", isi.Namespace, isi.Name, err))
+		return usage
+	}
+
+	// we cannot be sure, how many references will be added to the is, let's assume the worst case -
+	// we're going to reach the limit
+	specRefsIncrement.Set(int64(c.maximumTagsPerRepo))
+	allRefsIncrement.Set(int64(c.maximumTagsPerRepo))
+
+	// now let's decrease the istags increments for all the repository references already tagged in specs
+	// of all image streams in a project
+	for ref := range c.processedSpecRefs {
+		if specRefsIncrement.Value() <= 0 {
+			break
+		}
+		ref, parseErr := imageapi.ParseDockerImageReference(ref)
+		if parseErr != nil || ref.ID != "" {
+			continue
+		}
+		ref.Tag = ""
+		if ref.DaemonMinimal().Exact() == repoRef {
+			specRefsIncrement.Set(specRefsIncrement.Value() - 1)
+		}
+	}
+
+	// now let's decrease the isimages increments for all the repository references already tagged in a
+	// project
+	for ref := range c.processedRefs {
+		if allRefsIncrement.Value() <= 0 {
+			break
+		}
+		ref, parseErr := imageapi.ParseDockerImageReference(ref)
```

Comment:
So if I get your right, `imageapi.ResourceImageStreamImages` resource stands for a number of unique image hashes stored in statuses of all the project's image streams. Nothing in spec counts. All the references lacking image id are ignored. Please correct me if I'm wrong.

I like it much more actually. It's much easier to explain.